res/controllers/Pioneer-DDJ-SR2-scripts.js

```diff
@@ -0,0 +1,1383 @@
+
+    engine.beginTimer(300, function() {
+        midi.sendSysexMsg(DDJSR2.seratoControlsPoll, DDJSR2.seratoControlsPoll.length);
+    }, true);
+
+};
+
+DDJSR2.doSeratoHeartbeatTimer = function() {
+    midi.sendSysexMsg(DDJSR2.seratoHeartbeat, DDJSR2.seratoHeartbeat.length);
+};
+
+
+// First, a function library.
+////////////////
+
+
+DDJSR2.getRotaryDelta = function(value) {
+    return value > 0x40 ? value - 0x80: value;
+};
+
+DDJSR2.getJogWheelDelta = function(value) {
+    // The Wheel control centers on 0x40; find out how much it's moved by.
+    return value - 0x40;
+};
+
+DDJSR2.wheelLedControl = function(channelOffset, color) {
+    const wheelLedBaseChannel = 0xBB;
+    midi.sendShortMsg(
+        wheelLedBaseChannel,
+        0x20+(channelOffset),
+        color
+    );
+};
+
+
+DDJSR2.pitchBendFromJog = function(group, movement) {
+    engine.setValue(group, "jog", movement / 5 * DDJSR2.jogwheelSensitivity);
+};
+
+DDJSR2.RingBufferView = class {
+    constructor(indexable, startIndex = 0) {
+        this.indexable = indexable;
+        this.index = startIndex;
+    }
+    advanceBy(n) {
+        this.index = script.posMod(this.index + n, this.indexable.length);
+        return this.current();
+    }
+    next() {
+        return this.advanceBy(1);
+    }
+    previous() {
+        return this.advanceBy(-1);
+    }
+    current() {
+        return this.indexable[this.index];
+    }
+};
+
+DDJSR2.BrowserContainer = function() {
+    this.browseKnob  = new components.ComponentContainer({
+        longPressTimer: 0,
+        longPressTimeout: 250,
+        trackColorCycleEnabled: false,
+        trackColorCycleHappened: false,
+        previewSeekEnabled: false,
+        previewSeekHappened: false,
+        turn: new components.Encoder({
+            group: "[Library]",
+            inKey: "MoveVertical",
+            inValueScale: DDJSR2.getRotaryDelta
+        }),
+        press: new components.Button({
+            group: "[Library]",
+            key: "GoToItem"
+        }),
+        shiftPress: new components.Button({
+            group: "[Library]",
+            input: function(channel, control, value, _status, _group) {
+                if (!value) { // On release, not press. Allows for track color cycling.
+                    this.trackColorCycleEnabled = false;
+                    if (!this.trackColorCycleHappened) {
+                        if (!engine.getValue("[PreviewDeck1]", "play")) {
+                            engine.setValue("[PreviewDeck1]", "LoadSelectedTrackAndPlay", 1);
+                        } else {
+                            script.triggerControl("[PreviewDeck1]", "stop");
+                        }
+                    } else {
+                        this.trackColorCycleHappened = false;
+                    }
+                } else {
+                    this.trackColorCycleEnabled = true;
+                }
+            }.bind(this)
```

Comment:
deep nesting is usually a code smell (makes the code hard to follow). 
```suggestion
                
                if (this.isPress(channel, control, value, status, group)) {
                    theContainer.browseKnob.trackColorCycleEnabled = true;
                    return;
                }
                // On release, not press. Allows for track color cycling.
                theContainer.browseKnob.trackColorCycleEnabled = false;
                if (theContainer.browseKnob.trackColorCycleHappened) {
                    theContainer.browseKnob.trackColorCycleHappened = false;
                    return;
                }
                if (!engine.getValue("[PreviewDeck1]", "play")) {
                    engine.setValue("[PreviewDeck1]", "LoadSelectedTrackAndPlay", 1);
                } else {
                    script.triggerControl("[PreviewDeck1]", "stop");
                }
            }
```

Reply:
> where is theContainer defined? Do you mean DDJSR2.BrowserContainer? Should there be a theContainer = this; at the start of the BrowserContainer function, to define theContainer?

Yeah, you need to define it. I just couldn't add a suggestion for that because of github limitations.